disk: add a SDHC card over SPI driver.

ydamigos opened the pull request:
Add a SDHC card over SPI driver.

This PR is based on @nzmichaelh's work in #5528. I just updated to the new logger, disk access API and SPI API.

Tested on olimexino_stm32 using `samples/subsys/usb/mass`, "SDHC" as mass storage disk name and ELM FAT file system.

nzmichaelh committed:
disk: add a SDHC card over SPI driver.

Features:

- Uses the SPI bus to communicate with the card
- Detects and safely rejects SDSC (<= 2 GiB) cards
- Uses the optional CRC support for data integrity
- Retries resumable errors like CRC failure or temporary IO failure
- Works well with ELMFAT
- When used on a device with a FIFO or DMA, achieves >= 310 KiB/s on a
  4 MHz bus

Tested on a mix of SanDisk, Samsung, 4V, and ADATA cards from 4 GiB to
32 GiB.

Signed-off-by: Michael Hope <mlhx@google.com>

nzmichaelh committed:
lib: add crc7.

Used as a checksum on command messages when talking with MMC cards.

Implemented using the unwound bytewise implementation from
https://en.wikipedia.org/wiki/Computation_of_cyclic_redundancy_checks
which is a good mix of size and speed.

The API and naming matches lib/crc7.c in Linux.

Signed-off-by: Michael Hope <mlhx@google.com>

ydamigos committed:
boards/arm/olimexino_stm32: Add SDHC support

Add SDHC support over SPI2 port.

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

ydamigos committed:
disk: disk_access_sdhc: Move to new logger subsystem

Move to new logger subsystem.

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

ydamigos committed:
disk: disk_access_sdhc: Enable multiple disk interfaces

Enable multiple disk interfaces for SDHC.

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

ydamigos committed:
disk: disck_access_sdhc: Switch to new SPI API

Replace legacy API calls by news ones.

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

nashif commented on doc/subsystems/disk/sdhc.rst:
```diff
@@ -0,0 +1,56 @@
+adds SDHC card slot on `spi1`, uses `PA27` for chip select, and runs
+the SPI bus at 24 MHz once the SDHC card has been initialized.
+
+Using
```
Usage

dbkinder commented on doc/subsystems/disk/sdhc.rst:
```diff
@@ -0,0 +1,56 @@
+
+MMC and SDSC (<= 2 GiB) cards are not supported and will be ignored.
+
+The system does not support inserting or removing cards while the
```
This seems rather important to point out, so let's change to:
```
note:: The system does not support inserting or removing cards while the
   system is running.   The cards must be present at boot and must not be
   removed.  This may be fixed in future releases.
  
   FAT filesystems are not power safe and may become corrupted
   if power is lost or the card is removed.
```

dbkinder commented on doc/subsystems/disk/sdhc.rst:
```diff
@@ -0,0 +1,56 @@
+            };
+    };
+
+adds SDHC card slot on `spi1`, uses `PA27` for chip select, and runs
```
Let's move this before the code block rather than breaking the sentence into two parts.

dbkinder commented on doc/subsystems/disk/sdhc.rst:
```diff
@@ -0,0 +1,56 @@
+********
+
+The SDHC card slot is configured using Device Tree and should be added
+as a child of a SPI port.  For example, the fragment:
```
Change to:
```
For example, this device tree fragment adds an SDHC card slot on `spi1`,
uses `PA27` for chip select, and runs the SPI bus at 24 MHz once the
SDHC card has been initialized:
```

dbkinder commented on doc/subsystems/disk/sdhc.rst:
```diff
@@ -0,0 +1,56 @@
+The SDHC card will be automatically detected and initialized by the
+filesystem driver when the board boots.
+
+To read and write files and directories, see the file APIs in
```
Use a reference to the file system API doc:
```suggestion
To read and write files and directories, see the :ref:`file_system` in
```

dbkinder reviewed (changes requested):
Thanks for adding this subsystem documentation.

ydamigos commented:
@dbkinder I made the suggested changes.

dbkinder commented on doc/subsystems/disk/sdhc.rst:
```diff
@@ -0,0 +1,54 @@
+
+MMC and SDSC (<= 2 GiB) cards are not supported and will be ignored.
+
+note:: The system does not support inserting or removing cards while the
```
My bad... I left something off:
```suggestion
.. note:: The system does not support inserting or removing cards while the
```

ydamigos replied:
Fixed.

psidhu commented:
Is there any benefit into allowing this to be a bit more flexible to allow access via sdio vs spi? I can envision the answer to be "no" as there are no real sdio drivers in zephyr (other than the ext/hal stuff).

ydamigos commented:
> Is there any benefit into allowing this to be a bit more flexible to allow access via sdio vs spi? I can envision the answer to be "no" as there are no real sdio drivers in zephyr (other than the ext/hal stuff).

I just rebased and updated the original stalled PR to get it merged upstream. I don't have the time to make core changes to it.

dbkinder reviewed (approved):
doc changes LGTM, thanks!

nashif commented on include/crc7.h:
```diff
@@ -0,0 +1,36 @@
+/*
```
need to move this to the unified crc.h include file

ydamigos replied:
Done in b2375eeda10fef322fec20a68189bdc35ebd2539

ydamigos committed:
crc: Move crc7 into CRC single header

Move crc7 into CRC single header

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

Odjek commented:
Hello ! What you I have to change to make it work with SD/MMC (still with SPI ) ?
